src/halmos/calldata.py

```diff
@@ -210,3 +247,71 @@ def head_size(x):
         static = len(tails) == 0
 
         return EncodingResult(heads + tails, total_size, static)
+
+
+@dataclass(frozen=True)
+class FunctionInfo:
+    name: str | None = None
+    sig: str | None = None
+    selector: str | None = None
+
+
+def str_abi(item: dict) -> str:
+    def str_tuple(args: list) -> str:
+        ret = []
+        for arg in args:
+            typ = arg["type"]
+            match = re.search(r"^tuple((\[[0-9]*\])*)$", typ)
+            if match:
+                ret.append(str_tuple(arg["components"]) + match.group(1))
+            else:
+                ret.append(typ)
+        return "(" + ",".join(ret) + ")"
+
+    if item["type"] != "function":
+        raise ValueError(item)
+    return item["name"] + str_tuple(item["inputs"])
+
+
+def find_abi(abi: list, fun_info: FunctionInfo) -> dict:
```

Comment:
I haven't looked precisely, but this seems inefficient, no? We go from a function signature and iterate over abi items until we find one that matches? Is that because that's how the solc json output is stored?

Should we consider building our own index of funsig to funabi?

Reply:
agree. yes, abi items are given as list in json output.

btw, this is not new, but existing code moved here. improving it can be done in a separate pr. added todo comment for that.

Reply:
done